src/backend/vulkan/src/lib.rs

```diff
@@ -353,6 +354,205 @@ unsafe extern "system" fn debug_report_callback(
+        }
+        if cfg!(target_os = "windows") {
+            extensions.push(khr::Win32Surface::name());
+        }
+        if cfg!(target_os = "macos") {
+            extensions.push(ash::extensions::mvk::MacOSSurface::name());
+        }
+
+        extensions.push(ext::DebugUtils::name());
+        if cfg!(debug_assertions) {
+            #[allow(deprecated)]
+            extensions.push(ext::DebugReport::name());
+        }
+
+        extensions.push(vk::KhrGetPhysicalDeviceProperties2Fn::name());
+
+        // VK_KHR_storage_buffer_storage_class required for `Naga` on Vulkan 1.0 devices
+        if driver_api_version == Version::V1_0 {
+            extensions.push(vk::KhrStorageBufferStorageClassFn::name());
+        }
+
+        // Only keep available extensions.
+        extensions.retain(|&ext| {
+            if instance_extensions
+                .iter()
+                .find(|inst_ext| unsafe { CStr::from_ptr(inst_ext.extension_name.as_ptr()) == ext })
+                .is_some()
+            {
+                true
+            } else {
+                info!("Unable to find extension: {}", ext.to_string_lossy());
+                false
+            }
+        });
+        Ok(extensions)
+    }
+
+    pub fn required_layers<L>(
+        entry: &EntryCustom<L>,
+    ) -> Result<Vec<&'static CStr>, hal::UnsupportedBackend> {
+        let instance_layers = entry.enumerate_instance_layer_properties().map_err(|e| {
+            info!("Unable to enumerate instance layers: {:?}", e);
+            hal::UnsupportedBackend
+        })?;
+
+        // Check requested layers against the available layers
+        let mut layers: Vec<&'static CStr> = Vec::new();
+        if cfg!(debug_assertions) {
+            layers.push(CStr::from_bytes_with_nul(b"VK_LAYER_KHRONOS_validation\0").unwrap());
+        }
+
+        // Only keep available layers.
+        layers.retain(|&layer| {
+            if instance_layers
+                .iter()
+                .find(|inst_layer| unsafe {
+                    CStr::from_ptr(inst_layer.layer_name.as_ptr()) == layer
+                })
+                .is_some()
+            {
+                true
+            } else {
+                warn!("Unable to find layer: {}", layer.to_string_lossy());
+                false
+            }
+        });
+        Ok(layers)
+    }
+
+    /// # Safety
+    /// `raw_instance` must be created using at least the extensions provided by `Instance::required_extensions()`
+    /// and the layers provided by `Instance::required_extensions()`.
+    /// `driver_api_version` must match the version used to create `raw_instance`.
+    /// `extensions` must match the extensions used to create `raw_instance`.
+    /// `raw_instance` must be manually destroyed *after* gfx-hal Instance has been dropped.
+    pub unsafe fn from_raw(
+        #[cfg(not(feature = "use-rtld-next"))] entry: Entry,
+        #[cfg(feature = "use-rtld-next")] entry: EntryCustom<()>,
```

Comment:
we should probably use the fact that `type Entry = EntryCustom<Arc<Library>>`. So the only thing different between `use-rtld-next` and not is the generic parameter (which we can typedef somewhere).

Reply:
Great idea! Using cfg-gated parameters was also confusing rust-analyzer.

Reply:
I preferred using a typedef on `Entry`, typedef-ing only the type argument felt like splitting the type in half.